src/cdogs/keyboard.c

```diff
@@ -107,6 +108,93 @@ void KeyOnKeyUp(keyboard_t *keyboard, const SDL_Keysym s)
+    int currentTicks = (int)SDL_GetTicks(); //Used to determine whether to keep holding a diagonal by comparing how long it has been since it set that diagonal's ticks vs current time
+    //diagonal is held to give some leeway between releasing one half of it and the other so that it's easier to stop facing a diagonal
+    
+    const Uint8* realKeyboardState = SDL_GetKeyboardState(NULL); //This is the physical state of the keyboard untouched by any program manipulations, never manipulated, just compared against, eliminates a lot of bugs
+    
+    // -----
+    KeyPress currentPlayerKeys[4] = {(*keyboard).currentKeys[(*keyboard).PlayerKeys[c].up], (*keyboard).currentKeys[(*keyboard).PlayerKeys[c].right], 
+    (*keyboard).currentKeys[(*keyboard).PlayerKeys[c].down], (*keyboard).currentKeys[(*keyboard).PlayerKeys[c].left]}; //Needed an array of keyboard directions arranged in a circle to be able to urn this into a simple loop
+    
+    int realCurrentPlayerKeys[4] = {realKeyboardState[(*keyboard).PlayerKeys[c].up], realKeyboardState[(*keyboard).PlayerKeys[c].right],
+    realKeyboardState[(*keyboard).PlayerKeys[c].down], realKeyboardState[(*keyboard).PlayerKeys[c].left]};
+      //Needed a similar array for the actual state of the keyboard so 'i' would always match between currentKeys and realKeyboard
+    
+    //------
+    //If an opposing direction is pressed to what is currently sustained, immediately end sustain, an attempt to make sure false positives are nipped in the bud
+    for (int i = 0; i < 4; ++i)
+    {
+        if ((keyboard->diagonalStatus[c][i] == SUSTAIN) && ((currentPlayerKeys[(i + 2) % 4].isPressed == true || (currentPlayerKeys[(i + 3) % 4].isPressed == true))))
+            {
+                keyboard->diagonalStatus[c][i] = UNPRESSED;
+                keyboard->diagonalTicks[c][i] = (-1);
+                currentPlayerKeys[i].isPressed = realCurrentPlayerKeys[i];
+                currentPlayerKeys[(i + 1) % 4].isPressed = realCurrentPlayerKeys[(i + 1) % 4];
+            } 
+    }
+    
+    // -----
+    //Sets the stage to change to sustain, hopefully avoids false positives by creating an intermediary state between not pressed and a diagonal being sustained
+    for (int i = 0; i < 4; ++i) 
+    {
+        if ((currentPlayerKeys[i].isPressed == true && currentPlayerKeys[(i + 1) % 4].isPressed == true) && (keyboard->diagonalStatus[c][i] == UNPRESSED)) 
+            {
+                keyboard->diagonalStatus[c][i] = PRESSED;
+                keyboard->diagonalTicks[c][i] = (-1);
+            }
+    }
+    
+    // ------
+    //Sets sustain if both buttons in a diagonal have been pressed but one is no longer pressed 
+    for (int i = 0; i < 4; ++i) 
+    {
+        if ((keyboard->diagonalStatus[c][i] == PRESSED) && ((currentPlayerKeys[i].isPressed == true && currentPlayerKeys[(i + 1) % 4].isPressed == false)
+        || (currentPlayerKeys[i].isPressed == false && currentPlayerKeys[(i + 1) % 4].isPressed == true)))
```

Comment:
this can be simplified using `currentPlayerKeys[i].isPressed || currentPlayerKeys[(i + 1) % 4].isPressed`